File: src/main/java/com/example/finsentinel/service/okx/OkxApiClient.java
package com.example.finsentinel.service.okx;

import com.example.finsentinel.config.OkxProperties;
import com.example.finsentinel.service.okx.dto.*;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Service;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.Base64;
import java.util.List;
import java.util.Map;

/**
 * OKX v5 REST API client with HMAC-SHA256 authentication.
 *
 * <p>Uses Java 21's built-in {@link HttpClient} with HTTP/2. Every request is
 * signed per OKX's specification: {@code Base64(HMAC-SHA256(timestamp + METHOD
 * + requestPath + body, secretKey))} with four auth headers.
 *
 * <p>On any error the client returns {@code OkxResponse("1", errorMsg, [])}
 * instead of throwing, so callers can always pattern-match on {@code isSuccess()}.
 */
@Service
@Slf4j
@ConditionalOnProperty(name = "app.trading.okx.enabled", havingValue = "true")
public class OkxApiClient {

    private static final String HMAC_SHA256 = "HmacSHA256";
    private static final DateTimeFormatter TIMESTAMP_FMT =
            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
                    .withZone(ZoneOffset.UTC);

    private final OkxProperties properties;
    private final ObjectMapper objectMapper;
    private final HttpClient httpClient;

    public OkxApiClient(OkxProperties properties, ObjectMapper objectMapper) {
        this.properties = properties;
        this.objectMapper = objectMapper;
        this.httpClient = HttpClient.newBuilder()
                .version(HttpClient.Version.HTTP_2)
                .build();
    }

    // ── Account ─────────────────────────────────────────────────────────

    /** Get account balance across all currencies. */
    public OkxResponse<OkxAccountBalance> getBalance() {
        return get("/api/v5/account/balance", new TypeReference<>() {});
    }

    /** Get current positions. */
    public OkxResponse<OkxPosition> getPositions() {
        return get("/api/v5/account/positions", new TypeReference<>() {});
    }

    /** Get account configuration (trade mode, leverage, etc.). */
    public OkxResponse<Map<String, Object>> getAccountConfig() {
        return get("/api/v5/account/config", new TypeReference<>() {});
    }

    // ── Market ──────────────────────────────────────────────────────────

    /** Get ticker for a single instrument. */
    public OkxResponse<OkxTicker> getTicker(String instId) {
        return get("/api/v5/market/ticker?instId=" + instId, new TypeReference<>() {});
    }

    /** Get all tickers for an instrument type (SPOT, SWAP, FUTURES, OPTION). */
    public OkxResponse<OkxTicker> getTickers(String instType) {
        return get("/api/v5/market/tickers?instType=" + instType, new TypeReference<>() {});
    }

    /** Get current funding rate for a perpetual swap. */
    public OkxResponse<OkxFundingRate> getFundingRate(String instId) {
        return get("/api/v5/public/funding-rate?instId=" + instId, new TypeReference<>() {});
    }

    /** Get mark price for an instrument. */
    public OkxResponse<OkxTicker> getMarkPrice(String instId, String instType) {
        return get("/api/v5/public/mark-price?instId=" + instId + "&instType=" + instType,
                new TypeReference<>() {});
    }

    // ── Trade ───────────────────────────────────────────────────────────

    /** Place a new order. */
    public OkxResponse<OkxOrder> placeOrder(Map<String, Object> body) {
        return post("/api/v5/trade/order", body, new TypeReference<>() {});
    }

    /** Cancel an existing order. */
    public OkxResponse<OkxOrder> cancelOrder(String instId, String ordId) {
        return post("/api/v5/trade/cancel-order",
                Map.of("instId", instId, "ordId", ordId),
                new TypeReference<>() {});
    }

    /** Get pending (unfilled) orders. */
    public OkxResponse<OkxOrder> getPendingOrders() {
        return get("/api/v5/trade/orders-pending", new TypeReference<>() {});
    }

    /** Get order history for an instrument type. */
    public OkxResponse<OkxOrder> getOrderHistory(String instType) {
        return get("/api/v5/trade/orders-history-archive?instType=" + instType,
                new TypeReference<>() {});
    }

    // ── Margin ──────────────────────────────────────────────────────────

    /** Set leverage for an instrument. */
    public OkxResponse<Map<String, Object>> setLeverage(String instId, String lever, String mgnMode) {
        return post("/api/v5/account/set-leverage",
                Map.of("instId", instId, "lever", lever, "mgnMode", mgnMode),
                new TypeReference<>() {});
    }

    // ── Internal Helpers ────────────────────────────────────────────────

    /**
     * Signed GET request.
     */
    private <T> OkxResponse<T> get(String path, TypeReference<OkxResponse<T>> typeRef) {
        try {
            String timestamp = getTimestamp();
            String signature = sign(timestamp, "GET", path, "");

            HttpRequest.Builder builder = HttpRequest.newBuilder()
                    .uri(URI.create(properties.getBaseUrl() + path))
                    .GET()
                    .header("Content-Type", "application/json")
                    .header("OK-ACCESS-KEY", properties.getApiKey())
                    .header("OK-ACCESS-SIGN", signature)
                    .header("OK-ACCESS-TIMESTAMP", timestamp)
                    .header("OK-ACCESS-PASSPHRASE", properties.getPassphrase());

            if (properties.isSandbox()) {
                builder.header("x-simulated-trading", "1");
            }

            HttpResponse<String> response = httpClient.send(
                    builder.build(), HttpResponse.BodyHandlers.ofString());

            log.debug("OKX GET {} → {}", path, response.statusCode());
            return objectMapper.readValue(response.body(), typeRef);
        } catch (Exception e) {
            log.error("OKX GET {} failed: {}", path, e.getMessage());
            return new OkxResponse<>("1", e.getMessage(), List.of());
        }
    }

    /**
     * Signed POST request.
     */
    private <T> OkxResponse<T> post(String path, Object body, TypeReference<OkxResponse<T>> typeRef) {
        try {
            String bodyJson = objectMapper.writeValueAsString(body);
            String timestamp = getTimestamp();
            String signature = sign(timestamp, "POST", path, bodyJson);

            HttpRequest.Builder builder = HttpRequest.newBuilder()
                    .uri(URI.create(properties.getBaseUrl() + path))
                    .POST(HttpRequest.BodyPublishers.ofString(bodyJson))
                    .header("Content-Type", "application/json")
                    .header("OK-ACCESS-KEY", properties.getApiKey())
                    .header("OK-ACCESS-SIGN", signature)
                    .header("OK-ACCESS-TIMESTAMP", timestamp)
                    .header("OK-ACCESS-PASSPHRASE", properties.getPassphrase());

            if (properties.isSandbox()) {
                builder.header("x-simulated-trading", "1");
            }

            HttpResponse<String> response = httpClient.send(
                    builder.build(), HttpResponse.BodyHandlers.ofString());

            log.debug("OKX POST {} → {}", path, response.statusCode());
            return objectMapper.readValue(response.body(), typeRef);
        } catch (Exception e) {
            log.error("OKX POST {} failed: {}", path, e.getMessage());
            return new OkxResponse<>("1", e.getMessage(), List.of());
        }
    }

    /**
     * HMAC-SHA256 signature per OKX v5 specification.
     *
     * <p>Signature = Base64(HMAC-SHA256(timestamp + method + requestPath + body, secretKey))
     *
     * <p>Package-private for unit testing.
     */
    String sign(String timestamp, String method, String requestPath, String body) {
        try {
            String prehash = timestamp + method + requestPath + body;
            Mac mac = Mac.getInstance(HMAC_SHA256);
            SecretKeySpec keySpec = new SecretKeySpec(
                    properties.getSecretKey().getBytes(StandardCharsets.UTF_8), HMAC_SHA256);
            mac.init(keySpec);
            byte[] hash = mac.doFinal(prehash.getBytes(StandardCharsets.UTF_8));
            return Base64.getEncoder().encodeToString(hash);
        } catch (Exception e) {
            throw new IllegalStateException("Failed to compute HMAC-SHA256 signature", e);
        }
    }

    /**
     * ISO timestamp in UTC: {@code yyyy-MM-dd'T'HH:mm:ss.SSS'Z'}.
     *
     * <p>Package-private for unit testing.
     */
    String getTimestamp() {
        return TIMESTAMP_FMT.format(Instant.now());
    }
}

File: apps/api/src/okx/okx-api.client.ts
import { createHmac } from 'node:crypto';
import { Logger } from '@nestjs/common';
import type {
  OkxResponse,
  OkxTicker,
  OkxPosition,
  OkxOrder,
  OkxAccountBalance,
  OkxFundingRate,
} from './interfaces/okx-types';

/**
 * OKX REST API client with HMAC-SHA256 authentication.
 *
 * Sign: base64(hmacSha256(timestamp + method + path + body, secretKey))
 * Headers: OK-ACCESS-KEY, OK-ACCESS-SIGN, OK-ACCESS-TIMESTAMP, OK-ACCESS-PASSPHRASE
 * Optionally: x-simulated-trading: 1 for sandbox mode
 *
 * NOT an Injectable -- instantiated by OkxModule at runtime,
 * matching the pattern used by AlpacaTradingEngine.
 */
export class OkxApiClient {
  private static readonly DEFAULT_BASE_URL = 'https://www.okx.com';
  private readonly logger = new Logger(OkxApiClient.name);
  private readonly baseUrl: string;

  constructor(
    private readonly apiKey: string,
    private readonly secretKey: string,
    private readonly passphrase: string,
    baseUrl?: string,
    private readonly sandbox: boolean = false,
  ) {
    this.baseUrl = baseUrl ?? OkxApiClient.DEFAULT_BASE_URL;
  }

  // ---------------------------------------------------------------------------
  // Public API methods
  // ---------------------------------------------------------------------------

  /** GET /api/v5/market/ticker — fetch ticker for an instrument */
  async getTicker(instId: string): Promise<OkxTicker | null> {
    try {
      const response = await this.sendGet(
        `/api/v5/market/ticker?instId=${encodeURIComponent(instId)}`,
      );
      const body = (await response.json()) as OkxResponse<OkxTicker>;
      this.checkOkxResponse(body);
      return body.data[0] ?? null;
    } catch (err) {
      this.logger.error(`Failed to get ticker for ${instId}`, err);
      return null;
    }
  }

  /** GET /api/v5/account/balance — fetch account balance */
  async getAccountBalance(): Promise<OkxAccountBalance | null> {
    try {
      const response = await this.sendGet('/api/v5/account/balance');
      const body = (await response.json()) as OkxResponse<OkxAccountBalance>;
      this.checkOkxResponse(body);
      return body.data[0] ?? null;
    } catch (err) {
      this.logger.error('Failed to get account balance', err);
      return null;
    }
  }

  /** GET /api/v5/account/positions — fetch all positions */
  async getPositions(): Promise<OkxPosition[]> {
    try {
      const response = await this.sendGet('/api/v5/account/positions');
      const body = (await response.json()) as OkxResponse<OkxPosition>;
      this.checkOkxResponse(body);
      return body.data;
    } catch (err) {
      this.logger.error('Failed to get positions', err);
      return [];
    }
  }

  /** POST /api/v5/trade/order — place an order */
  async placeOrder(params: {
    instId: string;
    tdMode: string;         // "cross", "isolated", "cash"
    side: string;           // "buy" or "sell"
    ordType: string;        // "market", "limit"
    sz: string;             // size
    px?: string;            // price (for limit orders)
    reduceOnly?: boolean;
  }): Promise<OkxOrder | null> {
    try {
      const requestBody: Record<string, unknown> = {
        instId: params.instId,
        tdMode: params.tdMode,
        side: params.side,
        ordType: params.ordType,
        sz: params.sz,
      };

      if (params.px != null) {
        requestBody['px'] = params.px;
      }
      if (params.reduceOnly) {
        requestBody['reduceOnly'] = true;
      }

      const response = await this.sendPost('/api/v5/trade/order', requestBody);
      const body = (await response.json()) as OkxResponse<OkxOrder>;
      this.checkOkxResponse(body);
      return body.data[0] ?? null;
    } catch (err) {
      this.logger.error('Failed to place order', err);
      return null;
    }
  }

  /** GET /api/v5/trade/orders-pending — fetch all pending orders */
  async getPendingOrders(): Promise<OkxOrder[]> {
    try {
      const response = await this.sendGet('/api/v5/trade/orders-pending');
      const body = (await response.json()) as OkxResponse<OkxOrder>;
      this.checkOkxResponse(body);
      return body.data;
    } catch (err) {
      this.logger.error('Failed to get pending orders', err);
      return [];
    }
  }

  /** GET /api/v5/trade/orders-history — fetch historical orders */
  async getOrderHistory(instType?: string): Promise<OkxOrder[]> {
    try {
      let path = '/api/v5/trade/orders-history';
      if (instType) {
        path += `?instType=${encodeURIComponent(instType)}`;
      }
      const response = await this.sendGet(path);
      const body = (await response.json()) as OkxResponse<OkxOrder>;
      this.checkOkxResponse(body);
      return body.data;
    } catch (err) {
      this.logger.error('Failed to get order history', err);
      return [];
    }
  }

  /** POST /api/v5/trade/cancel-order — cancel an order */
  async cancelOrder(instId: string, ordId: string): Promise<boolean> {
    try {
      const response = await this.sendPost('/api/v5/trade/cancel-order', {
        instId,
        ordId,
      });
      const body = (await response.json()) as OkxResponse<unknown>;
      this.checkOkxResponse(body);
      return true;
    } catch (err) {
      this.logger.error(`Failed to cancel order ${ordId}`, err);
      return false;
    }
  }

  /** GET /api/v5/public/funding-rate — fetch funding rate for an instrument */
  async getFundingRate(instId: string): Promise<OkxFundingRate | null> {
    try {
      const response = await this.sendGet(
        `/api/v5/public/funding-rate?instId=${encodeURIComponent(instId)}`,
      );
      const body = (await response.json()) as OkxResponse<OkxFundingRate>;
      this.checkOkxResponse(body);
      return body.data[0] ?? null;
    } catch (err) {
      this.logger.error(`Failed to get funding rate for ${instId}`, err);
      return null;
    }
  }

  /** POST /api/v5/account/set-leverage — configure leverage for an instrument */
  async setLeverage(
    instId: string,
    leverage: string,
    marginMode: string,
  ): Promise<boolean> {
    try {
      const response = await this.sendPost('/api/v5/account/set-leverage', {
        instId,
        lever: leverage,
        mgnMode: marginMode,
      });
      const body = (await response.json()) as OkxResponse<unknown>;
      this.checkOkxResponse(body);
      return true;
    } catch (err) {
      this.logger.error(`Failed to set leverage for ${instId}`, err);
      return false;
    }
  }

  // ---------------------------------------------------------------------------
  // HMAC-SHA256 signature generation
  // ---------------------------------------------------------------------------

  /**
   * Generates the OKX API signature.
   * Sign = base64(hmacSha256(timestamp + method + requestPath + body, secretKey))
   */
  sign(
    timestamp: string,
    method: string,
    requestPath: string,
    body: string,
  ): string {
    const preSign = timestamp + method.toUpperCase() + requestPath + body;
    return createHmac('sha256', this.secretKey)
      .update(preSign)
      .digest('base64');
  }

  // ---------------------------------------------------------------------------
  // HTTP helpers
  // ---------------------------------------------------------------------------

  private authHeaders(
    method: string,
    requestPath: string,
    body: string,
  ): Record<string, string> {
    const timestamp = new Date().toISOString();
    const signature = this.sign(timestamp, method, requestPath, body);

    const headers: Record<string, string> = {
      'OK-ACCESS-KEY': this.apiKey,
      'OK-ACCESS-SIGN': signature,
      'OK-ACCESS-TIMESTAMP': timestamp,
      'OK-ACCESS-PASSPHRASE': this.passphrase,
      'Content-Type': 'application/json',
    };

    if (this.sandbox) {
      headers['x-simulated-trading'] = '1';
    }

    return headers;
  }

  private async sendGet(path: string): Promise<Response> {
    const headers = this.authHeaders('GET', path, '');
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'GET',
      headers,
    });
    this.checkHttpStatus(response);
    return response;
  }

  private async sendPost(
    path: string,
    body: Record<string, unknown>,
  ): Promise<Response> {
    const bodyStr = JSON.stringify(body);
    const headers = this.authHeaders('POST', path, bodyStr);
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers,
      body: bodyStr,
    });
    this.checkHttpStatus(response);
    return response;
  }

  private checkHttpStatus(response: Response): void {
    if (!response.ok) {
      throw new Error(
        `OKX API HTTP error ${response.status}: ${response.statusText}`,
      );
    }
  }

  private checkOkxResponse(response: OkxResponse<unknown>): void {
    if (response.code !== '0') {
      throw new Error(
        `OKX API error code ${response.code}: ${response.msg}`,
      );
    }
  }
}
